### ISSUE TITLE
Uncaught error with bad syntax

### ISSUE THREAD

**Issue by fubuloubu** [vyperlang/vyper#2116]
### Version Information

* vyper Version (output of `vyper --version`): 0.2.3
* OS: linux
* Python Version (output of `python --version`): 3.8.5

### What's your issue about?
This program fails to compile (as it should) with an uncaught error
```python
...
time_total: uint256 = self.auctions[_asset][_tokenId].deadline - time_elapsed, decimal
...
```
Stack:
```python
  File "brownie/_cli/__main__.py", line 56, in main
    importlib.import_module(f"brownie._cli.{cmd}").main()
  File "brownie/_cli/compile.py", line 37, in main
    project.load()
  File "brownie/project/main.py", line 656, in load
    return Project(name, project_path)
  File "brownie/project/main.py", line 160, in __init__
    self.load()
  File "brownie/project/main.py", line 213, in load
    self._compile(changed, self._compiler_config, False)
  File "brownie/project/main.py", line 87, in _compile
    build_json = compiler.compile_and_format(
  File "brownie/project/compiler/__init__.py", line 132, in compile_and_format
    output_json = compile_from_input_json(input_json, silent, allow_paths)
  File "brownie/project/compiler/__init__.py", line 243, in compile_from_input_json
    return vyper.compile_from_input_json(input_json, silent, allow_paths)
  File "brownie/project/compiler/vyper.py", line 55, in compile_from_input_json
    return vyper_json.compile_json(input_json, root_path=allow_paths)
  File "vyper/cli/vyper_json.py", line 392, in compile_json
    compiler_data, warn_data = compile_from_input_dict(input_dict, exc_handler, root_path)
  File "vyper/cli/vyper_json.py", line 324, in compile_from_input_dict
    return exc_handler(contract_path, exc, "compiler"), {}
  File "vyper/cli/vyper_json.py", line 100, in exc_handler_raises
    raise exception
  File "vyper/cli/vyper_json.py", line 316, in compile_from_input_dict
    data = vyper.compile_codes(
  File "vyper/opcodes.py", line 223, in _wrapper
    return fn(*args, **kwargs)
  File "vyper/compiler/__init__.py", line 110, in compile_codes
    raise exc
  File "vyper/compiler/__init__.py", line 105, in compile_codes
    out[contract_name][output_format] = OUTPUT_FORMATS[output_format](compiler_data)
  File "vyper/compiler/output.py", line 52, in build_abi_output
    abi = sig_utils.mk_full_signature(compiler_data.global_ctx)
  File "vyper/compiler/phases.py", line 90, in global_ctx
    self.vyper_module_folded, self.interface_codes
  File "vyper/compiler/phases.py", line 82, in vyper_module_folded
    validate_semantics(self._vyper_module_folded, self.interface_codes)
  File "vyper/context/validation/__init__.py", line 11, in validate_semantics
    validate_functions(vyper_ast)
  File "vyper/context/validation/local.py", line 59, in validate_functions
    FunctionNodeVisitor(vy_module, node, namespace)
  File "vyper/context/validation/local.py", line 166, in __init__
    self.visit(node)
  File "vyper/context/validation/base.py", line 19, in visit
    visitor_fn(node)
  File "vyper/context/validation/local.py", line 183, in visit_AnnAssign
    validate_expected_type(node.value, type_definition)
  File "vyper/context/validation/utils.py", line 325, in validate_expected_type
    given_types = _ExprTypeChecker().get_possible_types_from_node(node)
  File "vyper/context/validation/utils.py", line 70, in get_possible_types_from_node
    return fn(node)
  File "vyper/context/validation/utils.py", line 200, in types_from_Tuple
    return [TupleDefinition(types_list)]
  File "vyper/context/types/indexable/sequence.py", line 97, in __init__
    location = sorted((i.location for i in value_type), key=lambda k: k.value)[-1]
  File "vyper/context/types/indexable/sequence.py", line 97, in <genexpr>
    location = sorted((i.location for i in value_type), key=lambda k: k.value)[-1]
AttributeError: type object 'DecimalPrimitive' has no attribute 'location'
```

### How can it be fixed?
Unsure, seems like a weird syntax bug to do with Decimal types?

**Comment by iamdefinitelyahuman** [comment]
Took a look at this...  what you ran into is a small part of a much broader issue.  The type checker makes some assumptions that the values it's handed will be variables or literals, and so breaks in weird ways when it's handed things such as uncalled functions and type primitives.

PR incoming.

**Comment by fubuloubu** [comment]
The PR I just merged should have closed this, right?

**Comment by iamdefinitelyahuman** [comment]
Yes, fixed in #2118
